### ISSUE TITLE
tx.allow_html= is not defined as a part of the configuration

### ISSUE THREAD

**Issue by UncleIS** [SpiderLabs/owasp-modsecurity-crs#704]
Hello, community.

While debugging my configuration I noticed that rule 941150 was triggering. After looking into the code I found a stanza that declares that the rule should only trigger if setvar:tx.allow_html=0 is set somewhere, but I don't see a place where this var could be configured. 

Please, have a look at:

https://github.com/SpiderLabs/owasp-modsecurity-crs/blob/d1692bae42a0a1de81b074187250c71ded5a1d13/rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf#L202-L231

Thank you!

**Comment by dune73** [comment]
True. This seems to be an omission. Thank you for reporting.

As a workaround you can set the variable in your crs-setup.conf. Just copy one of the existing active SecActions, assign a new rule-id and enter the tx.allow_html variable.

**Comment by dune73** [comment]
Do you have a test-payload that should trigger this rule? Would be very convenient for me. :)

**Comment by UncleIS** [comment]
Unfortunately, message in error log was not created for some reason, but attached you will find audit and debug logs for the request that initiated the investigation. Hope that this will be helpful.

[archive.zip](https://github.com/SpiderLabs/owasp-modsecurity-crs/files/850755/archive.zip)

I tried to remove IP-addresses and hostnames as much as possible. Hopefully, this will not be a problem :)

Thank you!

**Comment by dune73** [comment]
Thank you. That looks good. 941150 triggered on the User-Agent  `Mozilla/5.0 (Windows NT 6.3; WOW64; Trident/7.0; Touch; rv:11.0) like Gecko`, but then the chained rule checking for allow_html failed because the variable is not existing.

It seems the resolution as for now is to 
- list this variable
- set it to 1, thus allowing html

@csanders-git @lifeforms @franbuehler : What do you think?

This is a PL1 rule and we overlooked the fact that the rule is disabled as is, but enabling it with allow_html = 0 would bring a ton of FPs. So if we want to fix it in PL1, we need to set it to 1.
For 3.1 we can look into shifting it to PL2 and maybe rework it a bit. They it is, there is a custom control just for this rule. Maybe we can streamline html detection generally. That could mean to remove the chained variable check and shift the rule to PL3.

**Comment by csanders-git** [comment]
@dune73, I think that suggestions makes a lot of sense. A PR to fix that would be easy, but we should also make a ticket so we don't forget to move things for 3.1

**Comment by franbuehler** [comment]
This is a nice finding!
Your suggestion sounds good, @dune73 :
For 3.0: List variable but set it to 0 to avoid the introduction of new FP.
For 3.1: Shift rule to PL2 (maybe revise the rule) and set variable to 1.

**Comment by dune73** [comment]
That you for your confirmation. I see that I messed up the values in my previous comment. So there is the plan:
```
3.0: tx.allow_html=1
3.1: tx.allow_html=0
```
I will also create a separate ticket to make sure we do not forget about this for 3.1.

**Comment by dune73** [comment]
- Fix for 3.0 is in PR #705.
- Description of planned fix for 3.1 is in issue #706.

**Comment by franbuehler** [comment]
I messed up the values, not you, sorry, @dune73.

**Comment by dune73** [comment]
The error was that I did not make it crystal clear from the start. ;)

**Comment by lifeforms** [comment]
CRS configuration used to contain the switches `tx.allow_html` and `tx.allowed_html_urls`.

If `tx.allow_html` was set, and the URL matched `tx.allowed_html_urls`, then XSS rules were skipped, allowing a user to post HTML.

If I remember correctly (though I cannot find my notes), we discussed that this feature was duplicate, since it is possible to selectively disable the XSS checks much more precisely using `SecRuleRemoveByTag` and `ruleRemoveTargetByTag` on HTML-containing parameters.

Therefore, we removed `tx.allow_html` in commit 79d3011b97331b56677701d0a8883c6e3269962a.

The remaining instance of it seems vestigial. Unfortunately, it seems due to me missing this, rule 941150 is currently not working now.

I would not re-introduce any form of `tx.allow_html`, since it is overlapping with our current exclusion mechanisms, but rather just remove references to it.

**Comment by dune73** [comment]
Good you remember all this, @lifeforms. It is all lost on me.

As the situation is, we have a broken functionality / rule in a stable release and we need to fix it. The most simple fix seems to be to reintroduce `tx.allow_html` for 3.0.1. But instead of shifting it to PL2/3 for 3.1 as proposed above, we better remove the functionality as you explain.

If we remove the chained check for `tx.allow_html` for 3.0.1, we introduce change the behaviour of the rule set fundamentally for PL1. So I do not think this is feasible. Removing the rule completely for 3.0.1 does not make much sense either, so I think my PR is still the best way to handle the present situation for 3.0.1.

**Comment by lifeforms** [comment]
If we remove the chained check, it means we will re-enable the rule 941150. (Rule 941150 is now unfortunately never firing in 3.0 since commit 79d3011, due to `TX:ALLOW_HTML` never being set.)

Do you mean that it's undesirable that the re-enabling rule 941150 could introduce additional false positives? That's true. In fact it might be a very good policy for us to never introduce new types of blockings in a bugfix version. That might encourage uptake of bugfix versions.

However, I still do not see a good reason to bring back `allow_html` functionality now (don't think anyone depends on it since we removed 99% of its functionality and only kept this effect by mistake) or in future. So why don't we just re-enable this rule in 3.1 by default then. But we could certainly put it in PL2 if it's FP prone, which might be discovered by testing.

**Comment by dune73** [comment]
Yes, I am quite sure a default-on 941150 brings too many FPs at PL1, so that's a no-go for 3.0.1.

But keeping an obviously broken rule around is hard to swallow. Re-introducing `tx.allow_html` would solve this for 3.0.1, but I agree it would also be a step back.

We could mark `tx.allow_html` in the crs-setup.conf as a temporary fix.

**Comment by dune73** [comment]
OK, this broken rule has just bitten me in the soft parts.

The problem is the setvars happen on the first rule, while the allow stuff is only checked in the chained rule. This results in a situation where anomaly score is raised, but no alert is written since the chained rule fails and log/auditlog is thus never triggered.

Consequently, I ended up with blocked requests on a production server without alerts being logged.

I conclude that this rule needs to be fixed for 3.0.1. We can not ship the broken rule and wait for 3.1 to re-enable it.

**Comment by lifeforms** [comment]
So the rule was active? That's great, then we are not raising FP by removing the chain right? :)

**Comment by dune73** [comment]
Haha. Are you trying to weasel out of this?

In fact yes, it is sort of active, but you do not notice it is active as it does not show up in the audit and the error log. it is only showing when you have a well-tuned system and still see blocked requests

The test run which turned these up had 1 FP in 35K requests on average on random data (encrypted payload arguments).

It is broken. It has to be fixed.

**Comment by lifeforms** [comment]
Weasel? If I'm correct, a possible argument against my unchanged position (keep the rule without the chain check) was invalidated, since right now the rule *is* already active. So removing the chain check cannot cause additional FP. I might be dense or misunderstanding the other options though.

**Comment by dune73** [comment]
Interesting philosophical questions surface here.

Is a rule active if it raises the score, but does not block (it has the `block` action defined) nor write an alert? I would argue that the setvar is active but the rule is only partially active.

Is it a false positive if the score is raised, but no alert message is written? Probably yes.

What will happen if we remove the `tx.allow_html`: FPs that have been invisible hitherto will become visible. We can then tell the user, that the rule has been technically active before, but they simply did not notice due to an omission on our behalf. I do not think users would like this.

I think we could remove the chained rule and push to PL2 immediately for 3.0.1. Is this really a smart move?

Personally, I still prefer a temporary fix via a default `tx.allow_html`.

**Comment by dune73** [comment]
We are a bit at a stalemate here @lifeforms and a stalemate won't get us past 3.0.1.

Do you have any new insight on the issue? I have nothing more to add and I think one of us needs to move to arrive at a solution.

**Comment by lifeforms** [comment]
Meh, I still really would hate reintroducing a config knob that we removed for a very good reason, and having users spend time looking at this change and their config, with no effect on blocking effectivity, just for the purpose of adding a blocking entry to the log in case this rule matches. Keeping the var check in this rule was a bug. We removed the var check for every other XSS rule. I see no value in it. If we don't get consensus, maybe our project lead can take the lead here?

**Comment by csanders-git** [comment]
haha, thanks @lifeforms . I think the best approach here is we introduce the 'knob' for 3.0.1 and mark it as a temporary fix for 3.0.x line that will be removed in 3.1

**Comment by dune73** [comment]
Thank you @csanders-git for making a call here.

Still, I thought some more about this and I came to the conclusion that shifting to 941150 for 3.0.1 and removing the chained `tx.allow_html` is a valid solution here. Here is why:
- We have FPs right now, they are just not visible. We remove the FPs from the default install and make them visible with this move. PL2 is arguably the right level for the rule (given the FPs we saw) so I think we do not need to linger any longer.
- Re-introducing the knob again is a step backwards. Removing `tx.allow_html` is a step forward and a definitive solution, while the re-introduction of the knob is only temporary.  
- Cohesion is important in a project. This issue and the re-introduction rubs @lifeforms the wrong direction. It is not important enough to warrant hard feelings. We have also shown that we are able to put our opinion aside and let somebody else take the decision. So I think we can now come together and solve this once and for all.

If you do not mind I will thus close PR #705 and reintroduce a new PR which removes the chained check and shifts the remaining rule to PL2.

**Comment by lifeforms** [comment]
No hard feelings, sorry if I sounded curt, might just be conversation style -- I just had a pretty strong opinion on this one issue, but it's also just a small issue. I am fine deferring to the majority. I like your proposal in your last post.

**Comment by dune73** [comment]
Please see PR #733.

**Comment by dune73** [comment]
Closing this after #733 and the final solution for this has been merged.
